worker/caasapplicationprovisioner/worker.go

```diff
@@ -163,20 +188,36 @@ func (p *provisioner) loop() error {
 				}
 
 				config := AppWorkerConfig{
-					Name:       app,
+					Name:       appName,
 					Facade:     p.facade,
 					Broker:     p.broker,
 					ModelTag:   p.modelTag,
 					Clock:      p.clock,
-					Logger:     p.logger.Child("applicationworker"),
+					Logger:     p.logger.Child(appName),
 					UnitFacade: p.unitFacade,
+					ShutDownCleanUpFunc: func() {
+						go p.shutDownAppWorker(appName)
+					},
 				}
 				startFunc := p.newAppWorker(config)
-				err = p.runner.StartWorker(app, startFunc)
+				p.logger.Debugf("starting app worker %q", appName)
+				err = p.runner.StartWorker(appName, startFunc)
 				if err != nil {
 					return errors.Trace(err)
 				}
 			}
 		}
 	}
 }
+
+func (p *provisioner) shutDownAppWorker(appName string) {
+	err := p.runner.StopAndRemoveWorker(appName, p.catacomb.Dying())
```

Comment:
We really should improve the runner to have a special error like ErrRemoveWorker

Reply:
yeah, we should consider having more specific errors for the runner.